source/common/upstream/cluster_manager_impl.cc

```diff
@@ -67,6 +67,11 @@ void ClusterManagerInitHelper::addCluster(Cluster& cluster) {
     cluster.initialize(initialize_cb);
   } else {
     ASSERT(cluster.initializePhase() == Cluster::InitializePhase::Secondary);
+    secondary_init_clusters_.remove_if(
+        [name_to_remove = cluster.info()->name()](Cluster* cluster_iter) {
+          return cluster_iter->info()->name() == name_to_remove;
+        });
```

Comment:
Why does this only apply to a secondary cluster? Couldn't a CDS update come in for a primary cluster and lead to the same problem? Either way please add more comments.

Reply:
My understanding is primary cluster list is determined by static config at ClusterManagerImpl constructor. There is no chance to replace primary cluster.
I could be wrong. I can add the remove attempt to  primary cluster if is needed

Reply:
I don't think this is true. Initialization looks like:
1) static primary
2) static secondary
3) CDS primary
3) CDS secondary

Check me on this, but assuming that is the case I think the issue can happen in both cases.

/wait

Reply:
I added the code in primary. But I haven't find the best to keep a primary cluster in warming and signal it ready so that I can prove primary clusters is unblocked.

Reply:
Think twice and I find the path can be covered by unit test.

Reply:
Unit test added @mattklein123 Could you take another look? Thanks